src/Utilities/TrackedDependencies/CanonicalTrackedInputFiles.cs

```diff
@@ -1114,15 +1047,9 @@ public void RemoveDependenciesFromEntryIfMissing(ITaskItem[] source, ITaskItem[]
             // Remove entries for each individual source
             for (int sourceIndex = 0; sourceIndex < source.Length; sourceIndex++)
             {
-                if (correspondingOutputs != null)
-                {
-                    rootingMarker = FileTracker.FormatRootingMarker(source[sourceIndex], correspondingOutputs[sourceIndex]);
-                }
-                else
-                {
-                    rootingMarker = FileTracker.FormatRootingMarker(source[sourceIndex]);
-                }
-
+                rootingMarker = correspondingOutputs != null
+                    ? FileTracker.FormatRootingMarker(source[sourceIndex], correspondingOutputs[sourceIndex])
+                    : FileTracker.FormatRootingMarker(source[sourceIndex]);
                 RemoveDependenciesFromEntryIfMissing(rootingMarker);
```

Comment:
I had to revert the change here.

The following code:
```
FileTracker.FormatRootingMarker(source[sourceIndex], correspondingOutputs?[sourceIndex])
```

Actually isn't the same. Passing null here actually calls an overload which wraps the value in an array, so we end up with `new[] { null }`. Then later code doesn't handle nulls inside the provided arrays very well.

Reply:
Ah. I looked at 

https://github.com/Microsoft/msbuild/blob/2a012e653766eb261b09b29b3106f4eb57a7f61d/src/Utilities/TrackedDependencies/FileTracker.cs#L322-L325

but it actually would call 

https://github.com/Microsoft/msbuild/blob/2a012e653766eb261b09b29b3106f4eb57a7f61d/src/Utilities/TrackedDependencies/FileTracker.cs#L311-L314

instead of this overload

https://github.com/Microsoft/msbuild/blob/2a012e653766eb261b09b29b3106f4eb57a7f61d/src/Utilities/TrackedDependencies/FileTracker.cs#L301-L304

Not confusing at all!